internal/agent/exercise.go

```diff
@@ -0,0 +1,82 @@
+package agent
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/aau-network-security/haaukins-agent/internal/environment/lab/exercise"
+	"github.com/aau-network-security/haaukins-agent/pkg/proto"
+	eproto "github.com/aau-network-security/haaukins-exercises/proto"
+	"github.com/rs/zerolog/log"
+)
+
+// GRPc endpoint that adds exercises to an already running lab. It requires the lab tag, and an array of exercise tags.
+// It starts by creating the containers needed for the exercise, then it refreshes the DNS and starts the containers afterwards.
+// It utilizes a mutex lock to make sure that if anyone tries to run the same GRPc call twice without the first being finished, the second one will wait
+func (a *Agent) AddExercisesToLab(ctx context.Context, req *proto.AddExerciseRequest) (*proto.StatusResponse, error) {
+
+	lab, err := a.State.EnvPool.GetLabByTag(req.LabTag)
+	lab.M.Lock()
+	defer lab.M.Unlock()
+
+	if err != nil {
+		log.Error().Str("labTag", req.LabTag).Err(err).Msg("error getting lab by tag")
+		return nil, err
+	}
+
+	var exerConfs []exercise.ExerciseConfig
+	exerDbConfs, err := a.State.ExClient.GetExerciseByTags(ctx, &eproto.GetExerciseByTagsRequest{Tag: req.Exercises})
+	if err != nil {
+		log.Error().Err(err).Msg("error getting exercise by tags")
+		return nil, errors.New(fmt.Sprintf("error getting exercises: %s", err))
+	}
+	//log.Debug().Msgf("challenges: %v", exerDbConfs)
+	// Unpack into exercise slice
+	for _, e := range exerDbConfs.Exercises {
+		ex, err := protobufToJson(e)
+		if err != nil {
+			return nil, err
+		}
+		estruct := exercise.ExerciseConfig{}
+		json.Unmarshal([]byte(ex), &estruct)
+		exerConfs = append(exerConfs, estruct)
+	}
+
+	// Add exercises to lab
+	ctx = context.Background()
```

Comment:
Use the context from the function instead :)

Reply:
Im actually thinking of just removing the context from these functions as it is not used.

However i experienced context errors when using the context from the GRPc call, and got no clue why